tools/toktx/image.hpp

```diff
@@ -187,7 +187,15 @@ class color<componentType, 4> : public color_base<componentType, 4> {
          if (i > 3) i = 3;
          comps[i] = val;
      }
-};
+     void normalize() {
+         auto len = r * r + g * g + b * b + a * a; // Not sure about a here
```

Comment:
Look at [renormalize_nornal_map](https://github.com/KhronosGroup/KTX-Software/blob/a7159924a6095488d8fb1ded358113d85813f669/lib/basisu/encoder/basisu_enc.h#L2649) in the BasisU encoder. Like astcencoder it works only with 4 component images. It does not include `a` in the calculation of length. which I'm sure is correct. In the output it uses the input value of `a` if the length is < something it calls validThreshold (0.077f). For the other case, I am surprised it doesn't crash as it is dereferencing index 3 in a Vec3F.

It  also is doing rounding and clamping which we likely need here too. If the component type is not float then we'll need to convert to float before doing this calculation I think.

Reply:
yea sounds reasonable. That bit of code has lots of defence against corner cases.

Reply:
@MarkCallow what are the acceptible types we need to consider in this normalisation. uint8, float32 are probably a must. Does it need to work with float16 too?

Reply:
unit8, uint16 & float32. Does the template class work with float16? If so we should do that one as well.

Reply:
[Here ](https://github.com/KhronosGroup/KTX-Software/pull/493/commits/4c00a183be6110593b92896671e99b01f2e3b9bf) is an implementation.

Reply:
Thanks for the fast work. It is going to take me more time than I have now to properly review the new code. Hopefully I can do it tomorrow evening my time.

In the meantime, why are you using basisu::clamp? There is a `cclamp` template already defined in image.hpp.

Reply:
fixed in https://github.com/KhronosGroup/KTX-Software/pull/493/commits/1376f772d49279ea4c594fa7214049b978960741